br/pkg/stream/rewrite_meta_rawkv.go

```diff
@@ -461,3 +469,318 @@ func (sr *SchemasReplace) RewriteKvEntry(e *kv.Entry, cf string) (*kv.Entry, err
+	newJobID, err := sr.genGenGlobalID(context.Background())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO: `ts` in table `gc_delete_range` can be job.BinlogInfo.FinishedTS+1
+	switch job.Type {
+	case model.ActionDropSchema:
+		var tableIDs []int64
+		if err := job.DecodeArgs(&tableIDs); err != nil {
+			return errors.Trace(err)
+		}
+		// Note: tableIDs contains partition ids, cannot directly use dbReplace.TableMap
+		/* TODO: use global ID replace map
+		 *
+		 *	for i := 0; i < len(tableIDs); i++ {
+		 *		tableReplace, exist := dbReplace.TableMap[tableIDs[i]]
+		 *		if !exist {
+		 *			return errors.Errorf("DropSchema: try to drop a non-existent table, missing oldTableID")
+		 *		}
+		 *		tableIDs[i] = tableReplace.NewTableID
+		 *	}
+		 */
+
+		argsSet := make(map[int64]struct{}, len(tableIDs))
+		for _, tableID := range tableIDs {
+			argsSet[tableID] = struct{}{}
+		}
+
+		newTableIDs := make([]int64, 0, len(tableIDs))
+		for tableID, tableReplace := range dbReplace.TableMap {
+			if _, exist := argsSet[tableID]; !exist {
+				return errors.Errorf("DropSchema: record a table, but it doesn't exist in job args")
+			}
+			newTableIDs = append(newTableIDs, tableReplace.NewTableID)
+			for partitionID, newPartitionID := range tableReplace.PartitionMap {
+				if _, exist := argsSet[partitionID]; !exist {
+					return errors.Errorf("DropSchema: record a partition, but it doesn't exist in job args")
+				}
+				newTableIDs = append(newTableIDs, newPartitionID)
+			}
+		}
+
+		if len(newTableIDs) != len(tableIDs) {
+			return errors.Errorf("DropSchema: try to drop a non-exsitent table/partition, whose oldID doesn't exist in tableReplace")
+		}
+
+		if err := insertDeleteRangeForTable(ctx, newJobID, newTableIDs, sr.RewriteTS); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	// Truncate will generates new id for table or partition, so ts can be large enough
+	case model.ActionDropTable, model.ActionTruncateTable:
+		tableReplace, exist := dbReplace.TableMap[oldTableID]
+		if !exist {
+			return errors.Errorf("DropTable/TruncateTable: try to drop a non-existent table, missing oldTableID")
+		}
+
+		// The startKey here is for compatibility with previous versions, old version did not endKey so don't have to deal with.
+		var startKey kv.Key // unused
+		var physicalTableIDs []int64
+		var ruleIDs []string // unused
+		if err := job.DecodeArgs(&startKey, &physicalTableIDs, &ruleIDs); err != nil {
+			return errors.Trace(err)
+		}
+		if len(physicalTableIDs) > 0 {
+			// delete partition id instead of table id
+			for i := 0; i < len(physicalTableIDs); i++ {
+				newPid, exist := tableReplace.PartitionMap[physicalTableIDs[i]]
+				if !exist {
+					return errors.Errorf("DropTable/TruncateTable: try to drop a non-existent table, missing oldPartitionID")
+				}
+				physicalTableIDs[i] = newPid
+			}
+			if err := insertDeleteRangeForTable(ctx, newJobID, physicalTableIDs, sr.RewriteTS); err != nil {
```

Comment:
is `sr.RewriteTS` can make sure that GC worker won't start load this task until we finish the point restore?

Reply:
sr.RewriteTS is the timestamp when restore starts. Generally, db starts gc these ranges in about 10 minutes.

gc will be done in advance, if restore takes more than 10 mins. In this case, rows will be moved to table `gc_delete_range_done` from table `gc_delete_range`, so that some kvs may be still retained. 😭

Reply:
insert rows into `gc_delete_range` after `RestoreKVFiles` is done.